File: src/main/java/com/truveta/opentoken/io/csv/PersonAttributesCSVWriter.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.csv;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Map;
import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVPrinter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.truveta.opentoken.io.PersonAttributesWriter;

/**
 * The PersonAttributeCSVWriter class is responsible for writing person
 * attributes to a CSV file.
 * It implements the {@link PersonAttributeWriter} interface.
 */
public class PersonAttributesCSVWriter implements PersonAttributesWriter {
    private static final Logger logger = LoggerFactory.getLogger(PersonAttributesCSVWriter.class);

    private final BufferedWriter fileWriter;
    private final CSVPrinter csvPrinter;
    private final String filePath;
    private boolean headerWritten = false;

    /**
     * Initialize the class with the output file in CSV format.
     * 
     * @param filePath the output file path
     * @throws IOException if an I/O error occurs
     */
    public PersonAttributesCSVWriter(String filePath) throws IOException {
        this.filePath = filePath;
        fileWriter = new BufferedWriter(new FileWriter(filePath));
        csvPrinter = new CSVPrinter(fileWriter, CSVFormat.DEFAULT);
    }

    @Override
    public void writeAttributes(Map<String, String> data) {

        try {
            if (!headerWritten) {
                // Write the header
                csvPrinter.printRecord(data.keySet());
                headerWritten = true;
            }

            csvPrinter.printRecord(data.values());

        } catch (IOException e) {
            logger.error("Error in writing CSV file: {}", e.getMessage());
        }
    }

    @Override
    public void close() throws Exception {
        this.csvPrinter.close();
        this.fileWriter.close();
    }
}

File: lib/nodejs/opentoken/src/io/csv/PersonAttributesCSVWriter.ts
/**
 * Copyright (c) Truveta. All rights reserved.
 */

import * as fs from 'fs';
import { stringify } from 'csv-stringify';
import { PersonAttributesWriter } from '../PersonAttributesWriter';

/**
 * Writes person attributes to a CSV file.
 */
export class PersonAttributesCSVWriter implements PersonAttributesWriter {
  private stringifier: any;
  private stream: fs.WriteStream;

  constructor(filePath: string) {
    this.stream = fs.createWriteStream(filePath);
    this.stringifier = stringify({
      header: true,
    });
    this.stringifier.pipe(this.stream);
  }

  async writeAttributes(data: Map<string, string>): Promise<void> {
    const record: Record<string, string> = {};
    for (const [key, value] of data.entries()) {
      record[key] = value;
    }

    return new Promise((resolve, reject) => {
      this.stringifier.write(record, (err: Error | null) => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
    });
  }

  async close(): Promise<void> {
    return new Promise((resolve) => {
      this.stringifier.end(() => {
        this.stream.end(resolve);
      });
    });
  }
}
